### PR TITLE
rate limits documentation

### PR DESCRIPTION
**Pull request by himalb001**
Public documentation of GraphQL resource limits.

More context:
https://3.basecamp.com/3453178/buckets/1933834/messages/6221093655
https://3.basecamp.com/3453178/buckets/23328660/messages/6161786100#__recording_6176276293

This PR 
* adds a nav item under API > Graphql > Limits
* Adds markdown file with graphql rate limits documentation

### CODE REVIEW

File: pages/apis/graphql/graphql_resource_limits.md
Comment (yob) [inline pages/apis/graphql/graphql_resource_limits.md:136]:
```suggestion
RateLimit-Limit: 20000
```

**Review by JuanitoFatas — COMMENTED** [review]
😍

File: pages/apis/graphql/graphql_resource_limits.md
Comment (shevaun) [inline pages/apis/graphql/graphql_resource_limits.md:111]:
```suggestion
To ensure optimal performance, an organization can use up to 20,000 actual complexity points within a 5 minute period. By allowing a set number of actual complexity points, we provide organizations flexibility to run queries of different sizes within a 5 minute window.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (shevaun) [inline pages/apis/graphql/graphql_resource_limits.md:113]:
```suggestion
As a best practice, we recommend utilizing client size strategies like caching to lower the number of API calls, queues to schedule API calls or proper use of pagination to only request necessary data to manage time based rate limits.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (shevaun) [inline pages/apis/graphql/graphql_resource_limits.md:169]:
```suggestion
* Optimize the request by only requesting data your app requires. We recommend using specific queries rather than a single all-purpose query.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (shevaun) [inline pages/apis/graphql/graphql_resource_limits.md:170]:
```suggestion
* Always use appropriate `first` or `last` values when requesting connections. Not providing those may default to 500 which can increase the requested complexity exponentially.
```?

Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md]: Ah, yes - that's true. Some of the newer connection that uses graphql-ruby can set default connection size

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:5]: Is there meant to be an on in here?

```suggestion
Our limits are based on the complexity of the query. We calculate complexity based on requested resources of the query and encourage developers to responsibly use techniques for limiting calls, pagination, caching and retrying requests to lower the complexity of the queries.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:3]:
```suggestion
To ensure that Buildkite stays stable for everyone, we have set some limits on our GraphQL API. These limits prevent excessive or abusive calls to our servers while still allowing API users to use GraphQL endpoints in a wide range of ways.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:36]:
```suggestion
Although these default costs are in place, Buildkite reserves the right to set manual costs to individual fields.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:44]:
```suggestion
The requested complexity is calculated based on the number of fields and objects requested. Usually, requesting a deeply nested query or excluding pagination details from connections results in high requested complexity.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:63]:
```suggestion
The actual complexity is based on the results returned after the query execution, since the connection fields can return fewer nodes than requested. Lowering the requested complexity usually lowers the actual complexity of queries.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:86]:
```suggestion
Buildkite has implemented two distinct limits to our GraphQL endpoints. These limits play a critical role in ensuring that our platform operates smoothly and efficiently, while minimising the risk of unnecessary downtime or system failures.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (jameshill) [inline pages/apis/graphql/graphql_resource_limits.md:111]: I know we've talked about it before, but with a requested complexity limit of 50K and an Actual of 20K over 5 minutes, the first query can use 50K points and then subsequent queries will fail, is that correct?

**Review by jameshill — COMMENTED** [review]
This is such a good read!!!

not sure if my suggestions are good... just throwing them in there to confirm my thinking.

💚 @himalb001

Just waiting on a review from @mbelton-buildkite and then we'll need to talk logistics agin before releasing!

Talk tomorrow!

Final thought.... there's quite a few double spaced lines.. do we try to only have blank line between blocks given markdown strips them anyway?

File: pages/apis/graphql/graphql_resource_limits.md
Comment (yob) [inline pages/apis/graphql/graphql_resource_limits.md:142]:
```suggestion
`RateLimit-Reset` is the number of seconds remaining until a new time window is started and the limits are reset.  
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:111]: Yup!

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:171]: General tightening:

```suggestion
* Use strategies like caching for data you use often that is unlikely to be updated instead of constantly calling APIs.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:172]: Use active voice:

```suggestion
* Regulate the rate of your requests for smoother distribution. You can do this using queues or scheduling API calls in appropriate intervals.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:173]: Avoiding "app" + US spelling:

```suggestion
* Use metadata about your API usage, including rate limit status to manage the behavior dynamically.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:174]: Use serial comma:

```suggestion
* Think of rate limiting while designing your client application. Be mindful of retries, errors, loops, and the frequency of API calls.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:43]: Markdown style:

```suggestion
### Requested complexity

```

**Review by mbelton-buildkite — COMMENTED** [review]
I've left a lot of comments here but there shouldn't be anything major. This is really great content 👏🏻

Let me know if you want to chat about any of it 😊

File: data/nav_graphql.yml
Comment (himalb001) [reply data/nav_graphql.yml:9]: Good point! 

My reasoning was that people may not want to know about limits straight after reading graphql overview.
Maybe they go through some tutorials and examples and then they're presented with our limits?

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md]: Good point! I like the suggestion!

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:34]: I had that initially but the table width looked funny to me. I'm happy to merge the change though.

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:83]: Ah nice catch. Probably artefacts from the draft

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:44]: I think the second option makes more sense

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:143]: Good point - What about `Accessing time based limits details` and `Viewing query complexity`

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:107]: Probably by both - We render numbers without commas in API responses in few other places so it should be okay here.

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:138]: How about `Check time based limits` and `View query complexity`?

**Comment by himalb001** [comment]
Hey @mbelton-buildkite, I've resolved most of the comments. What do you think?

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:138]: Typo: Missing dash

```suggestion
### Check-time based limits
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:174]: Typo:

```suggestion
Consider the following best practices when designing your API usage:
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:176]: Missing "the":

```suggestion
- Optimize the request by only requesting the data you require. We recommend using specific queries rather than a single all-purpose query.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:144]: This is listing values and their descriptions, so it should use list formatting:

```suggestion
- `RateLimit-Remaining` — The remaining complexity left within the current time window.  
- `RateLimit-Limit` — The complexity limit for the time window.  
- `RateLimit-Reset` — The number of seconds remaining until a new time window is started and the limits are reset.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:145]: I think this should call out that you're about to view example values:

```suggestion

For example:

```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [inline pages/apis/graphql/graphql_resource_limits.md:134]: Remove unnecessary blank line:

```suggestion
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [reply pages/apis/graphql/graphql_resource_limits.md:139]: It would be good to have a sentence between these headings. Is there something you think would fit well there?

**Review by mbelton-buildkite — COMMENTED** [review]
Just a few more tweaks and it should be good to go 😊

File: pages/apis/graphql/graphql_resource_limits.md
Comment (JuanitoFatas) [reply pages/apis/graphql/graphql_resource_limits.md:138]: Should the dash for the time-based instead? 😅

```suggestion
### Check time-based limits
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (himalb001) [reply pages/apis/graphql/graphql_resource_limits.md:140]: What do you think of this?

```
You can access both time-based limits and query complexity information through the APIs. Accessing these limit details will not incur any additional complexity points.
```

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [reply pages/apis/graphql/graphql_resource_limits.md]: 🤦🏻‍♂️ Yes, thank you!

File: pages/apis/graphql/graphql_resource_limits.md
Comment (mbelton-buildkite) [reply pages/apis/graphql/graphql_resource_limits.md:140]: Yip, that's great! Nice context about not costing additional complexity points.

I think s/APIs/API, though.

**Review by mbelton-buildkite — APPROVED** [review]
Great work on this! I think it's my favorite new page. 👏🏻

**Pull request merged**
